### ISSUE TITLE
Qwen-3-Omni SFT and RL training slowdown with Megatron after upgrading to vLLM ≥ 0.12 (audio-in-video enabled)

### ISSUE THREAD

**Issue by callmeluoq-art** [modelscope/ms-swift#7108]
Background

I am training Qwen3-Omni using Megatron-LM, including both SFT and GRPO (RLHF) workflows.

To enable audio-in-video support (use_audio_in_video) required by Qwen3-Omni, I upgraded vLLM to ≥ 0.12, where this feature is introduced.

⸻

Issue

After upgrading the environment, I observed a severe training throughput regression, affecting both SFT and GRPO runs.
	•	Training becomes approximately 10–20× slower compared to the previous setup.

⸻

Environment comparison

Hardware:
- GPU:  H200

Previous (fast):
- CUDA: 12.8 
- PyTorch: 2.9.0+cu130 
- vLLM 0.9.3.dev6+gca66cbff0.precompiled

Current (slow, required for audio-in-video):
- CUDA: 13.0 (slow)
- PyTorch: 2.9.0+cu130 
- vLLM 0.12.1.dev0+g4fd9d6a85.d20251217.cu130

Training Script

megatron sft \
    --model /path/to/Qwen3-Omni-30B-A3B-Instruct \
    --dataset /path/to/dataset.jsonl \
    --train_type full \
    --attn_impl flash_attention_3 \
    --attention_backend flash \
    --tensor_model_parallel_size 4 \
    --expert_model_parallel_size 4 \
    --pipeline_model_parallel_size 1 \
    --context_parallel_size 1 \
    --global_batch_size 64 \
    --micro_batch_size 1 \
    --max_length 16384 \
    --max_epochs 1 \
    --lr 1e-5 \
    --min_lr 1e-6 \
    --lr_warmup_fraction 0.03 \
    --lr_decay_style cosine \
    --bf16 true \
    --freeze_vit true \
    --freeze_aligner true \
    --freeze_llm false \
    --sequence_parallel true \
    --packing false \
    --split_dataset_ratio 0.01 \
    --vit_gradient_checkpointing true \
    --moe_permute_fusion true \
    --moe_grouped_gemm true \
    --moe_shared_expert_overlap true \
    --moe_aux_loss_coeff 1e-3 \
    --cross_entropy_loss_fusion true \
    --load_safetensors true \
    --save_safetensors true \
    --load_from_cache_file true \
    --lazy_tokenize true \
    --save_interval 200 \
    --eval_interval 100 \
    --log_interval 1 \
    --log_throughput true \
    --log_timers_to_tensorboard true \
    --log_memory_to_tensorboard true \
    --log_params_norm true \
    --log_completions true \
    --no_save_optim \
    --no_save_rng \
    --num_workers 32 \
    --dataset_num_proc 8 \
    --wandb_project megatron_swift \
    --finetune true \
    --megatron_extra_kwargs '{"timing_log_level": 2}'

megatron rlhf \
    --rlhf_type grpo \
    --model /path/to/Qwen3-Omni-30B-A3B-Instruct \
    --dataset /path/to/dataset.jsonl#8000 \
    --train_type full \
    --tensor_model_parallel_size 4 \
    --expert_model_parallel_size 4 \
    --pipeline_model_parallel_size 1 \
    --context_parallel_size 1 \
    --global_batch_size 1024 \
    --micro_batch_size 1 \
    --num_generations 8 \
    --steps_per_generation 1 \
    --max_length 16384 \
    --max_completion_length 2048 \
    --max_pixels 1003520 \
    --lr 1e-6 \
    --min_lr 1e-7 \
    --lr_warmup_fraction 0.05 \
    --bf16 true \
    --beta 0.00 \
    --epsilon 3e-4 \
    --epsilon_high 4e-4 \
    --loss_type grpo \
    --importance_sampling_level sequence \
    --dynamic_sample false \
    --overlong_filter true \
    --use_vllm true \
    --vllm_mode colocate \
    --vllm_tensor_parallel_size 8 \
    --vllm_gpu_memory_utilization 0.4 \
    --vllm_max_model_len 16384 \
    --vllm_mm_processor_cache_gb 0 \
    --reward_funcs random_reward \
    --reward_weights 1.0 \
    --external_plugins /path/to/plugin.py \
    --freeze_vit true \
    --freeze_aligner true \
    --freeze_parameters talker code2wav \
    --attention_backend flash \
    --sequence_parallel true \
    --padding_free true \
    --offload_model true \
    --use_precision_aware_optimizer \
    --recompute_granularity selective \
    --sleep_level 2 \
    --moe_permute_fusion true \
    --moe_grouped_gemm true \
    --moe_shared_expert_overlap true \
    --moe_aux_loss_coeff 1e-3 \
    --vit_gradient_checkpointing false \
    --save_interval 50 \
    --eval_interval 50 \
    --log_interval 1 \
    --log_completions true \
    --no_save_optim true \
    --no_save_rng true \
    --num_workers 8 \
    --dataset_num_proc 8 \
    --wandb_project megatron_swift \
    --finetune \
    --seed 42

**Question**
Is there a recommended way to avoid the severe performance regression?

**Comment by Jintao-Huang** [comment]
use torch2.8

https://github.com/modelscope/ms-swift/issues/6744

**Comment by callmeluoq-art** [comment]
Since vLLM 0.12 is the first version to officially support passing the use audio in video–related arguments in vLLM Engine v1 for Qwen3-Omni, and vLLM 0.12 depends on PyTorch 2.9,
if I need to continue using PyTorch 2.8, do I have to downgrade vLLM to 0.9.3?

Additionally, if vLLM is downgraded to 0.9.3, does the Megatron GRPO training pipeline still support passing and correctly applying the use audio in video parameter?

**Comment by Jintao-Huang** [comment]
PyTorch 2.9 has a bug where conv3d runs very slowly. Could you search online for any solutions?

**Comment by callmeluoq-art** [comment]
I did some investigation on this and here is my current understanding, which may help explain the slowdown.

After profiling the training workflow, it seems that the major overhead comes from Conv3D in the ViT patch embedding of Qwen3-Omni.
In PyTorch 2.9, Conv3D appears to be significantly slower than in previous versions, and this might be the reason for the large throughput regression I observed during both SFT and GRPO training.

Based on this observation, I tried a local workaround on my side.

Specifically, during initialization I replace the original Conv3D-based forward in
Qwen3OmniMoeVisionPatchEmbed with an unfold + matmul (F.linear) implementation (similar to what vLLM uses).
The patch is applied only when running with PyTorch 2.9, so it should not affect other setups.

Below is the exact code I added in swift/megatron/__init__.py:


`import torch
import torch.nn.functional as F

def _apply_conv3d_patch():
    """Apply Conv3d patch for PyTorch 2.9 to avoid performance regression."""
    if not torch.__version__.startswith("2.9"):
        return

    try:
        from transformers.models.qwen3_omni_moe import modeling_qwen3_omni_moe

        def _patched_forward(self, x):
            """Use matrix multiplication instead of conv3d (reference: vLLM implementation)."""
            L, C = x.shape
            x = x.view(L, -1, self.temporal_patch_size, self.patch_size, self.patch_size)
            B, C, T, H, W = x.shape
            K = self.proj.kernel_size
            x = x.unfold(2, K[0], K[0]).unfold(3, K[1], K[1]).unfold(4, K[2], K[2])
            x = x.permute(0, 2, 3, 4, 1, 5, 6, 7).reshape(-1, C * K[0] * K[1] * K[2])
            x = F.linear(x, self.proj.weight.view(self.embed_dim, -1), self.proj.bias)
            return x.view(L, self.embed_dim)

        modeling_qwen3_omni_moe.Qwen3OmniMoeVisionPatchEmbed.forward = _patched_forward
        print("✅ [ms-swift] Patched Qwen3OmniMoeVisionPatchEmbed.forward for PyTorch 2.9 conv3d workaround")
    except ImportError:
        pass  # transformers version may not have this model
    except Exception as e:
        print(f"⚠️ [ms-swift] Failed to apply conv3d patch: {e}")

_apply_conv3d_patch()`


With this workaround, the training speed on my side improves significantly. Is my understanding and modification correct?

**Comment by Jintao-Huang** [comment]
https://github.com/modelscope/ms-swift/pull/7122
